hydroflow/src/compiled/pull/symmetric_hash_join.rs

```diff
@@ -1,24 +1,53 @@
 use std::collections::HashMap;
 
 #[derive(Debug)]
-pub struct JoinState<K, V1, V2> {
-    ltab: HashMap<K, Vec<V1>>,
-    rtab: HashMap<K, Vec<V2>>,
-    lbuffer: Option<(K, V1, Vec<V2>)>,
-    rbuffer: Option<(K, V2, Vec<V1>)>,
+pub struct HalfJoinState<K, VBuild, VProbe> {
+    table: HashMap<K, Vec<VBuild>>,
+    buffer: Option<(K, VProbe, usize)>,
 }
-
-impl<K, V1, V2> Default for JoinState<K, V1, V2> {
+impl<K, VBuild, VProbe> Default for HalfJoinState<K, VBuild, VProbe> {
     fn default() -> Self {
         Self {
-            ltab: HashMap::new(),
-            rtab: HashMap::new(),
-            lbuffer: None,
-            rbuffer: None,
+            table: HashMap::new(),
+            buffer: None,
+        }
+    }
+}
+impl<K, VBuild, VProbe> HalfJoinState<K, VBuild, VProbe>
+where
+    K: Clone + Eq + std::hash::Hash,
+    VBuild: Clone + Eq,
+    VProbe: Clone,
+{
+    fn pop_buffer(&mut self) -> Option<(K, VProbe, VBuild)> {
+        let (k, v, idx) = self.buffer.as_mut()?;
+        let row = &self.table[k];
+        let result = (k.clone(), v.clone(), row[*idx].clone());
```

Comment:
How deep is this clone? When does this get called -- every probe? Ideally use an `Rc` or `Cow` for all the data in the dataflow. If this is too much to fix now, please make this a perf issue to explore later.

Reply:
This is a bit of an improvement of before, before the entire matching "row" (vec) was cloned. now we don't make duplicate vecs, we just clone out of the original vec. Should perform better but needs testing